### ISSUE TITLE
Support for row-level security (RLS)

### ISSUE THREAD

**Issue by wladiston** [prisma/prisma#12735]
## Problem

Currently, if we want to guarantee that user's requests have enough permissions to run a query, we have to keep coding `where:{...} ` clauses everywhere in our code. Depending on the size of the code base, it can be extremely hard to make sure that all query conditions are correct, updated and concise with the data model.

## Suggested solution
We could have a way to implement some sort of `@rls` directly on the `schema.prisma` file. 
Something like this:

```prisma
-- defines what should be sent throght 
define session {
  id         String
  role       Role
}

enum Role {
  USER
  ADMIN
}

model Cart {
  id         String     @id @default(uuid())
  customer   Profile   @relation(fields: [customerId], references: [id])
  customerId String    
	
  @rls(read: ({row, session}) => row.customerId === session.id || session.role === ADMIN)
}
```

The migration tool would then make sure that this RLS rule would be synced with the database and the prisma client could make sure that the session is provided

```ts
await prisma.cart.findMany({
  define: {session},
  data: {...},
})

await prisma.cart.update({
  define: {session:{role: 'ANON'}}, // Only owner & ADMINs can update
  data: {...},
})

await prisma.$transaction([
  ...,
  await prisma.cart.findMany({
    // define: {session}, should not exist
    data: {...},
  })
], {define: {session})
```

## Alternatives

As suggested by @psugihara, the RLS could be implemented 'virtually' within prisma for unsupported databases. So maybe instead of having the migration tool to generate the SQL, we could have some sort of `prisma.rls.ts` file with all CRUD functions and in the `@rls` we just point it to the function we want to use to validade the query.

```prisma
model Cart {
 ...	
  @rls(read: "onlyAdmins")
}
```
```ts
export async function onlyAdmins(row, session){
  return session.role === "ADMIN"
}
```

## Additional context
I've tried to implement it off prisma and ended up making a full report of the adventure here: https://github.com/prisma/prisma/issues/5128#issuecomment-1059814093
I'd extremely encourage going over all points I've made.

**Comment by janpio** [comment]
Can you provide some sample SQL the Migration tooling would need to create for the hypothetical schema you provided above? Thanks.

**Comment by sergiocarneiro** [comment]
This would be really useful, and I'm also in favor of code-splitting, as most of the rules will be repetitive.

Here's my naming proposal, to be more descriptive and paradigm-agnostic:
```prisma
model Cart {
 ...	
 @@security(read: "ownerOrAdmins")
}
```

**Comment by wladiston** [comment]
Good point @sergiocarneiro. Allowing them be more descriptive could improve readability for those cases
"teamMemberWithPermisisonToReadOrAdmin" as well

**Comment by wladiston** [comment]
> Can you provide some sample SQL the Migration tooling would need to create for the hypothetical schema you provided above? Thanks.

Sure, I've updated both alternatives with the migration script

**Comment by Liam-Tait** [comment]
What if the new security attribute worked more like a `where` query.

I think this would be beneficial as:
- the schema maintains source of truth
- developers use a querying system they already know

Some rules for use:
- `context security` must be defined to use `@@security` otherwise schema is invalid
- `@@security` attribute must have a name
- `context: {security}` can be added to queries, must match `context security` structure

Using the above examples and converting to this approach. It would look like:

```prisma
enum Role {
  USER
  ADMIN
}

context security {
  id         String
  role       Role
}

model Cart {
  cartId     String     @id @default(uuid())
  customer   Profile   @relation(fields: [customerId], references: [id])
  customerId String    

  @@security(name: 'customer_cart', read: { id: { equals: customerId }})
  @@security(name: 'admin_cart', all: { role: { equals: ADMIN }})
}
```
This would generate for [PostgreSQL](https://www.postgresql.org/docs/current/sql-createpolicy.html)
```sql
-- Add Security Policy
CREATE POLICY "customer_cart" on "Cart" for select using ( session().id = customerId);

-- Add Security Policy
CREATE POLICY "admin_cart" on "Cart" for ALL using ( session().role = "ADMIN"  );
```

When querying, a context with security can optionally be added
```ts
const security = { id: 'the customer id', role: 'USER' }
await prisma.cart.findMany({ context: { security }})
```

**Comment by Liam-Tait** [comment]
I think it is a requirement to go from context to row, this is because if starting from the row it is not possible for use cases such as role matching. Which imo will be the most common situation.


```prisma
// Super user can do anything
@@security(name: 'superuser_cart', all: { role: { equals: SUPERUSER }})
// Admin can read anything
@@security(name: 'admin_cart', read: { role: { equals: ADMIN }})
// User can do anything on their own cart
@@security(name: 'user_cart', all: { role: { equals: USER }, id: { equals: customerId })
```

Treating the security context more like a model and allowing relations there could help solve this.
```prisma
context security {
  id   String
  user User   @relation(fields: [id], references: userId)

  role Role
}
```
The security for "User can access carts from Users in the same team" could then reference the context's user

```prisma
enum Role {
  USER
  ADMIN
}

context security {
  id   String
  user User   @relation(fields: [id], references: userId)

  role Role
}

model Team {
  teamId String @id @default(uuid())
  User   User[]
}

model User {
  userId String @id @default(uuid())

  teamId   String
  team     Team       @relation(fields: [teamId], references: [teamId])
  cart     Cart[]
}

model Cart {
  cartId String @id @default(uuid())

  userId String
  user   User?  @relation(fields: [userId], references: [userId])
  
  // A user can access carts created by users on the same team
  @@security(
    name: 'team_cart', 
    read: { user: { team: { teamId: { equals: user.teamId }}}}}
  )
}
```

**Comment by wladiston** [comment]
I see. Both ways are actually valid. I kinda don't feel comfortable doing a channelled reference of some property in the context because there's no consistency with the rest of the schema file.
Maybe cart is not the best example to represent all possible use cases.

I think we can let developers decided which way they want go and something like this would work amazingly well:

```prisma
enum Role {
    USER
    ADMIN
}

enum Permission {
    EDITOR
    ADMIN
}

context {
    // in this context, the props represent the user request
    id   String
    role Role
    teams String[] // so that we can use IN operators as well? but that would require to save the list of the teams in the jwt or something
}

model Team {
    teamId String @id @default(uuid())
    members   MemberOf[]
    websites   Website[]
}

model MemberOf {
    id String @id @default(uuid())
    teamId String
    team   Team  @relation(fields: [teamId], references: [teamId])

    userId String
    user   User  @relation(fields: [userId], references: [userId])

    permission Permission
}

model User {
    userId String @id @default(uuid())
    teams  MemberOf[]
    role Role
}

model Website {
    cartId String @id @default(uuid())

    teamId String
    team   Team?  @relation(fields: [teamId], references: [teamId])

    // it would match props from context
    @@security(
        name: 'Team members can view websites',
        read: { row: { teamId: { IN: [teams] } } }
    )

    // or, for nested row props
    // @@security(
    //     name: 'Team members can view websites',
    //     read: { row: { team: { members: { userId: [id] } } } }
    // )

     // to match values from context
    @@security(
        name: 'Admins can do anything to websites',
        all: { context: { role: ADMIN } }
    )
    
    @@security(
        name: 'Only team admins can delete websites',
        delete: { row: { team: { members: { userId: [id], permission: ADMIN } } } }
    )

    // @@security(
    //     name: 'Some random usecase that uses row AND context',
    //     create: { 
    //         row: { ... },
    //         context: { ... },
    //     }
    // )
    // @@security(
    //     name: 'Some random usecase that uses row OR context',
    //     update: { 
    //         OR: [
    //             { row: { ... } },
    //             { context: { ... } },
    //         ]
    //     }
    // )
}
```

Schema files could get immensely long though. At least it wouldn't have to have a different file to handle permissions and since it would use the same "where" structure, transforming it in a SQL or making it virtually would be simple. I'm just not sure how the Prisma vscode plugin would handle to have autocomplete before we `prisma generate` the new structure

**Comment by eduhenke** [comment]
I'd just like to provide a complex policy that we've described in our system, in case it helps designing with these use-cases in mind(SELECTing from other tables, transforming the data from the context, arbitrary conditional expressions):

```sql
CREATE POLICY station_policy ON "Station" USING (
  -- See only your organization's stations.
  "orgId" = any(string_to_array(current_setting('app.current_organizations'), ','))
  AND EXISTS (
      SELECT * FROM "Agreement"
      WHERE (
          "Station"."id" = "Agreement"."stationId"
          AND "Agreement"."agreementId" = any(string_to_array(current_setting('app.current_agreements', true), ','))
          AND "Agreement"."active" = true
      ))
  AND (
    -- Evaluate whether the requester is a member
    (current_setting('app.current_user_id', true)::int = -1)
    -- Non-exclusive stations
    OR (cardinality("groupIds"::text []) = 0)
    -- Get a station that has a common group with the requester
    OR (cardinality(current_setting('app.current_groups')::text []) <> 0 AND "groupIds" && current_setting('app.current_groups')::text [])
  )
);
```

We're passing the variables via the `current_setting` on PostgreSQL.

**Comment by mime29** [comment]
Thanks a lot for the work on this feature. Is there a way to set a milestone for this? I'm asking because we are considering dropping Prisma in our project if RLS is not supported in a close future. If the feature is to be added, we will simply keep Prisma with our explicit where condition related to the `user_id` until we can update our code base. The reason behind this is because we want to rely on the configuration more than on our unit tests when it comes to security.
Thank you in advance for the updates on this topic.

**Comment by janpio** [comment]
I recently dug into PostgreSQL RLS a bit and now understand the discussion above a lot better than before.

The suggested `@@security` (or variants) for `CREATE POLICY` is pretty straightforward, but will be _really_ hard to get right as you already identified - but solvable. No big unknowns there, just needs a really thorough design to cover all the things possible in SQL (which usually are... many).

Where I am currently still a bit unsure is how queries optimally get the "context" set, to make sure all queries are properly covered (so no way to accidentally not have a context set when running a query - and then failing) and reset (so no way to leak some context to another query that does not set its own context) and what is the _correct_/best way to actually run the SQL queries. Is it really as simple as running `SET LOCAL` before each query? How do other ORMs or database clients handle this, especially if they maintain a connection pool as Prisma does?

**Comment by andyjy** [comment]
> no way to leak some context to another query [that does not set its own context]

AFAIK yes - the only way to achieve this (assuming connection pooling) is wrapping context inside transactions using `SET LOCAL`

> no way to accidentally not have a context set when running a query

I'm currently using the dependency injection pattern to pass request-specific context[^1] to a wapping function that I use around all my prisma calls. This works, but isn't super pleasing developer experience. I'm optimistic about refactoring to [use a Proxy object around Prisma](https://github.com/prisma/prisma/issues/5128#issuecomment-1270602183) instead of the wrapper, and potentially [using AsyncLocalStorage instead of dependency injection](https://nodejs.org/docs/latest-v16.x/api/async_context.html#class-asynclocalstorage) to pass the context from my request handlers.

Assuming this approach works(!), having native support for e.g. `prismaClient.setRequestContext(..)` that uses `AsyncLocalStorage` internally to scope to the current async call stack (i.e. current request only) and handles the transaction-wrapping + `SET LOCAL` could feasibly make life a lot easier in the future in terms of passing the context for each query. _(Or potentially as a [Client extension](https://github.com/prisma/prisma/issues/15074#issuecomment-1282419200) as recently proposed - haven't investigated the feasibility of this yet.)_

> How do other ORMs or database clients handle this?

Frankly I haven't discovered native RLS functionality in other ORMs. So also interested in answers to this! It may be fair to say Prisma would be paving new ground if it added native support for RLS vs. having to custom-build. 

A useful reference point could be noting Supabase's approach - essentially just providing their request context within Postgres via a couple helper functions, leaving the developer to build everything else with raw SQL: https://supabase.com/docs/guides/auth/row-level-security

[^1]: for me: `user_id`, `tenant_id` for multi-tenant, also app-specific current `workspace_id`

**Comment by janpio** [comment]
(We are talking to Supabase about this as well, and currently looking in how their API layer (PostgREST, which the Supabase JS SDK talks to) implements this under the hood.)

**Comment by SorenHolstHansen** [comment]
One approach, and I don't know how well this fits with prisma as a whole, is the solution described in [this article](https://aws.amazon.com/blogs/database/multi-tenant-data-isolation-with-postgresql-row-level-security/) from aws SaaS factory, and specifically the "Alternative approach" part, though it says there are some problems with PgBouncer and similar. 

The basic idea is to inject the user into a database session when you get a pool connection. I know prisma does not allow for getting a connection and working with that, instead only working on a "Pool level", but if it could be possible to just do something like
```
const connection = await prisma.getConnection();
connection.$executeRaw(`SET app.current_tenant = '${TenantContext.getTenant()}'`);

// Fetch, or do whatever you want with the connection
connection.myRLSModel.findMany(...) // No need for any context in here, just do a regular query
```
It would enable RLS immediately. The connection would have the same API as the prisma client, but it would ensure that the function using the connection owns that connection, and within that connection, the user is verified.

I don't know how difficult this would be to pull off, but it seems somewhat easier than the above, with a context.

It would still require the `@@security` attribute to define the RLS on the model

**Comment by janpio** [comment]
I think the "Alternative Approach" is actually what we are mostly talking about, and others found ways to implement via middleware and similar before. For Prisma that raises exactly the problem, that we use a pool of connections and require to be able to use any free connection. As soon as we start pinning connections to a specific user (or web session etc), the performance behavior of Prisma changes completely and most users would run out of database connections. To counteract that, you have to reset each connection before or after using it - which is what we are trying to find a good way to implement.

**Comment by Liam-Tait** [comment]
If Prisma supported setting runtime configuration parameters, it would make it easier to implement row level security without Prisma needing to directly support row level security as a feature.

This could look like:
```ts
await prisma.post.findUniqueOrThrow({
  where: { id: 1 },
  local: { accountId: 2 }
)
```

If local is provided, the query would be a transaction and before the query add `SET LOCAL` for each key in local e.g `SET LOCAL accountId TO 2`

This solution is probably useful (not for me, I'm guessing) for other situations too such as setting the timezone, setting the date style for a query, because it is not tied directly to row level security.

I use PostgreSQL so that's what I am familiar with, but a quick look shows that this should work similarly in MySQL and MariaDb too but not SqLite

A simplified version of our current solution looks something like the following.
```js
const transactionAsUser = async (args,{ prisma, user }) => {
  const results = await prisma.$transaction([
    prisma.$executeRaw`SET LOCAL current_user_id TO ${user.id}::int)`,
    ...args
  ])
  return results.slice(1)
}

await transactionAsUser([
    prisma.post.findUniqueOrThrow({ where: { id: 1 } })
  ],
  { prisma, user }
)
```

**Comment by janpio** [comment]
Before starting with middlewares, maybe take a look at Prisma Client Extensions (preview): https://www.prisma.io/docs/concepts/components/prisma-client/client-extensions The `query` component is probably a better way to do similar things: https://www.prisma.io/docs/concepts/components/prisma-client/client-extensions/query

**Comment by kachar** [comment]
@MoSattler I think it's a good practice to add `$before` and `$after` statements to clear the value that we set in the beginning of the transaction. 

If we don't do that we might end up having the same `current_user_id` for multiple users that share a connection.

**Comment by kachar** [comment]
@MoSattler Alright, just confirmed the docs - indeed it says the effect is only while the transaction is active.

> The effects of SET LOCAL last only till the end of the current transaction, whether committed or not. A special case is SET followed by SET LOCAL within a single transaction: the SET LOCAL value will be seen until the end of the transaction, but afterwards (if the transaction is committed) the SET value will take effect.

Thanks for pointing that out.

**Comment by jamesmeneghello** [comment]
@MoSattler I think the use of `await prisma.$transaction` in your example is problematic? Since it would be accessing the original client, not the extended client. 

This is what I came to using NestJS / Prisma:

Use `async-hooks` to provide the org/tenant id for the request:

`request.context`:
```typescript
import { RequestContext } from '@medibloc/nestjs-request-context';

export class APIRequestContext extends RequestContext {
  orgId: number;
}
```

`main.ts`:
```typescript
import { RequestContext, requestContextMiddleware } from '@medibloc/nestjs-request-context';
import { APIRequestContext } from './util/request.context';
...
async function bootstrap() {
...
app.use(requestContextMiddleware(APIRequestContext));

// use a global middleware to attach the tenant to the async context (first in line after request handling)
app.use((req: Request, res: Response, next: NextFunction) => {
        const ctx: APIRequestContext = RequestContext.get();

        // just setting a static orgId, but pull the org from `req`, header or whatever you desire
        ctx.orgId = 1;
        
        next();
        
        // make sure to unset it, just in case
        ctx.orgId = undefined;
    });
...
}
```

`prisma.service.ts`:
```
import { RequestContext } from '@medibloc/nestjs-request-context';
import { INestApplication, Injectable, OnModuleInit } from '@nestjs/common';
import { Prisma, prisma, PrismaClient } from '@prisma/client';
import { APIRequestContext } from './util/request.context';

const authedClient = Prisma.defineExtension((client) => {
    const ctx: APIRequestContext = RequestContext.get();
    const orgId = ctx.orgId.toString();
    
    return client.$extends({query: {
        $allModels: {
          async $allOperations({ args, query }) {
            const [,result] = await client.$transaction([
              client.$executeRaw`select set_config('user.userId',${orgId},true)`,
              query(args)
            ]);
            return result;
          },
        },
      },
    });
})

@Injectable()
export class PrismaService extends PrismaClient implements OnModuleInit {

    constructor() {
        super({log: ['query', 'info', 'warn', 'error']});
    }
    
  async onModuleInit() {
    await this.$connect();
  }

  authed() {
    return this.$extends(authedClient);
  }

  async enableShutdownHooks(app: INestApplication) {
    this.$on('beforeExit', async () => {
      await app.close();
    });
  }
}
```

`posts.service.ts`:
```typescript
@Injectable()
export class PostsService {
    constructor(private prisma: PrismaService) {}

    ...
    findAll() {
        return this.prisma.authed().posts.findMany();
    }
    ...
}
```

The model:
```prisma
model Posts {
  id  Int   @default(autoincrement()) @id
  content String
  orgId Int
}
```

In a migration to set up the `posts` model:

```sql
ALTER TABLE "Posts" ENABLE ROW LEVEL SECURITY;
ALTER TABLE "Posts" FORCE ROW LEVEL SECURITY;

CREATE POLICY tenant_isolation_policy ON "Posts"
USING (
    "Posts"."orgId"::text = current_setting('user.userId', true)::text   
);

CREATE POLICY admin_isolation_policy ON "Posts"
USING (
    current_setting('user.admin', true) = 'true'
);
```

Doing this also meant that `seed` was unable to function (creating rows belonging to other organisations would fail using an RLS-obeying user, which our pg user is by default. Hence, the override if `user.admin` is set.

`seed.ts`:
```typescript
import { PrismaClient } from '@prisma/client'
const prisma = new PrismaClient()
async function main() {
    // for this seed, ignore RLS
    // you could also turn it off/on, but there are plenty of scenarios from within your API
    // that you'd want to freely access the tables
    await prisma.$executeRaw`select set_config('user.admin','true',false)`;

    for (let i = 0; i < 10; i++) {
        await prisma.posts.create({
            data: {
                content: "this is a post for org 1",
                orgId: i+1
            }
        })
    }
}
main()
  .then(async () => {
    await prisma.$disconnect()
  })
  .catch(async (e) => {
    console.error(e)
    await prisma.$disconnect()
    process.exit(1)
  })
```

For allowing free access to tables, we can then also create an `unauthed` client extension in the service that sets `user.admin` to `true`, or check it against the user's organisation's attributes, or etc.

Works really well though, thanks for all the ideas in the thread!

**Comment by jamesmeneghello** [comment]
Actually, one issue I'm now handling is that any query errors are instead thrown as client extension errors, e.g.:

```
PrismaClientExtensionError: Error caused by an extension: 
          Invalid `this.prisma.authed().posts.update()` invocation in
          /src/nestjs-template/api/src/posts/posts.service.ts:23:43
          
            20 }
            21 
            22 update(id: number, updatePostDto: UpdatePostDto) {
          → 23     return this.prisma.authed().posts.update(
          An operation failed because it depends on one or more records that were required but not found. Record to update not found.
```

**Comment by itspers** [comment]
@jamesmeneghello this looks finally right, though i still not sure if all this headache with RLS worth it - all this hacking prisma client, writing 2 policies for each table, and then need to un-hack it back for seeding, and we also have not only seeding but whole second app for managing tenants.. 

Im still thinking of if this has sense to implement, or just to write lint rule to enforce each prisma where to contain tenantId

Btw how update-create queries works with RLS, you still need to pass orgId there in code?

**Comment by jamesmeneghello** [comment]
I mean, that's shared schema multitenancy for you, really. A bit of pain here, or pain on the deployment end when deploying an environment-per-tenant (or pain when migrating schema-per-tenant), or etc. Much like everything in IT, you're dealing with sets of trade-offs, not a single obvious solution. I don't think this is too bad, considering.

Update queries work without passing orgId as part of the where - it just checks the existing orgId value on the row to be updated. Creates would require you to supply an orgId.

**Comment by itspers** [comment]
@jamesmeneghello yeah, but i started it few months ago, saw this issue and decided to postpone this problem. Now i have same nestjs setup, just without RLS - all controllers have :orgId in path, they validate it, pass to methods, methods to services, services insert it to prisma calls.

And if i will add RLS on top of it - i will just have more code, problems with seeding and extra cpu-work by postgres of processing all this RLS and set_config..  And all 'security' it adds - ability to write rare queries with findMany({})...

**Comment by shoooe** [comment]
@jamesmeneghello @MoSattler Doesn't wrapping every query in a transaction make it impossible to use transactions in application code? Moreover isn't that costly performance wise?

**Comment by jamesmeneghello** [comment]
@shoooe Postgres, at least, supports nested transactions - so it shouldn't interfere with any wrapping transactions you're using at a higher level. I haven't used this code in production yet, so haven't tested it for performance. I'm about to work it into a project though, so I'll test it out and let you know. It would likely be more performant to keep a request-scoped copy of the Prisma client and configure the user scoping there, but I haven't used Prisma enough to work out the fine details on whether that's feasible or not.

**Comment by shoooe** [comment]
@jamesmeneghello In the meantime I've found out that actually every statement is already wrapped in a transaction by Postgres so explicit transactions should always either improve performance or keep it the same.

> PostgreSQL actually treats every SQL statement as being executed within a transaction. If you do not issue a BEGIN command, then each individual statement has an implicit BEGIN and (if successful) COMMIT wrapped around it.

Reference: https://www.postgresql.org/docs/current/tutorial-transactions.html

**Comment by jiashengguo** [comment]
The Prisma extension library [ZenStack](https://zenstack.dev/?utm_campaign=github&utm_medium=organic&utm_content=prisma_rls) we are building now happens to share the same view with @wladiston and @Liam-Tait's solution to use the schema as the single source of truth. 🤝

This is how the sample schema should be represented in ZenStack now: 

```prisma
enum Role {
    USER
    ADMIN
}

model Cart {
    id String @id @default(uuid())
    customer Profile @relation(fields: [customerId], references: [id])
    customerId String

    //Carts are only visible by owners
    @@allow('read', customerId == auth().id)

    //Admins can do anything with carts
    @@allow('all', auth().role == 'ADMIN')
}
```

The only differences are:

1. We use @@allow and @@deny instead of  @@security
2. We don’t generate RLS for now.  Instead, we add a transparent proxy on Prisma client. 

For the multitenancy issues, we invited a [Collection Predict Expression](https://zenstack.dev/docs/reference/zmodel-language#collection-predicate-expressions?utm_campaign=github&utm_medium=organic&utm_content=prisma_rls) to express the policy rules for more complex "to-many" relations. Here is how it looks now:

```prisma
enum Role {
    USER
    ADMIN
}

model User {
    id            String    @id @default(cuid())
    spaces SpaceUser[]
    
    // can be created by anyone, even not logged in
    @@allow('create', true)

    // can be read by users sharing any space
    @@allow('read', spaces?[space.members?[user == auth()]])

    // full access by oneself
    @@allow('all', auth() == this)
}

model Space {
    id String @id @default(uuid())
    members SpaceUser[]

    // require login
    @@deny('all', auth() == null)

    // everyone can create a space
    @@allow('create', true)

    // any user in the space can read the space
    @@allow('read', members?[user == auth()])

    // space admin can update and delete
    @@allow('update,delete', members?[user == auth() && role == ADMIN])
}

/*
 * Model representing membership of a user in a space
 */
model SpaceUser {
    id String @id @default(uuid())
    space Space @relation(fields:[spaceId], references: [id], onDelete: Cascade)
    spaceId String
    user User @relation(fields: [userId], references: [id], onDelete: Cascade)
    userId String
    role Role

    // require login
    @@deny('all', auth() == null)

    // space admin can create/update/delete
    @@allow('create,update,delete', space.members?[user == auth() && role == ADMIN])

    // user can read entries for spaces which he's a member of
    @@allow('read', space.members?[user == auth()])
}
```

If you feel interested, there is a tutorial post for the complete version I wrote about it:
[How to build a collaborative SaaS product using Next.js and ZenStack's access control policy](https://zenstack.dev/blog/saas-demo?utm_campaign=github&utm_medium=organic&utm_content=prisma_rls)

We would really appreciate it if you could share your opinions by commenting or joining our [Discord](https://go.zenstack.dev/chat) to help us make [ZenStack](https://zenstack.dev/?utm_campaign=github&utm_medium=organic&utm_content=prisma_rls) the right thing to solve your problems.

**Comment by moraisp** [comment]
One problem that I found is that by just wrapping everything into it's own transaction we loose the[ automatic batch ](https://www.prisma.io/docs/guides/performance-and-optimization/query-optimization-performance) functionality prisma offers.

To go around that I build a simple throttling logic using lodash (you shouldn't need it if you don't want).

It's far from perfect and there are a few things that I'm not completely happy with so I'm up to suggestions on how to improve this code. But it is working pretty consistently for me.

```typescript
const transactionClient = client.$extends({
  name: 'transaction-client',
  client: {
    $transaction(arg, options) {
      if (Array.isArray(arg)) {
        const { before, after } = getRlsPromises(client); // build yours
        return client
          .$transaction([...before, ...arg, ...after] as [...P])
          .then((results) => results.slice(before.length, before.length + arg.length));
      } else {
        return client.$transaction(async (prisma) => {
          const { before, after } = getRlsPromises(prisma); // build yours
          await Promise.all(before);
          const res = await arg(prisma);
          await Promise.all(after);
          return res;
        }, options);
      }
    },
  },
});

const promiseMap = new Map<
  string,
  { query: (...args) => Promise<any>; args: any; resolve: (...args) => void; reject: (...args) => void }[]
>();

const throtleMap = new Map<string, (key: string) => void>();
const getThrottle = () =>
  _.throttle(
    (key) => {
      const value = promiseMap.get(key);
      promiseMap.delete(key);
      transactionClient
        .$transaction(() => {
          return Promise.allSettled(
            value.map(({ query, args, resolve, reject }) => {
              return query(args).then(resolve).catch(reject);
            })
          );
        })
        .catch(console.error.bind(null, 'transaction error'));
    },
    10, // this number should really be zero, I'm not sure why there are so many calls coming in different ticks in my app
    { leading: false }
  );

const rlsClient = transactionClient.$extends({
  name: 'rls-client',
  query: {
    $allModels: {
      $allOperations: ({ model, operation, args, query }) => {
        // I remove those operations from the batching logic since I dont want "mutations" to be implicitly batched together
       // maybe only findUnique instead of $allOperations
        if (['create', 'delete', 'update', 'deleteMany', 'updateMany', 'upsert'].includes(operation)) {
          return transactionClient.$transaction([query(args)]).then((res) => res[0]);
        }

        // I group then by key cause Prisma batching was very pick with the order things come by when it comes to batching
        const key = `${model}|${operation}`;
        return new Promise((resolve, reject) => {
          const throttle = throtleMap.get(key) ?? getThrottle();
          throtleMap.set(key, throttle);
          const value = promiseMap.get(key) ?? [];
          promiseMap.set(key, [...value, { query, args, resolve, reject }]);
          throttle(key);
        });
      },
    },
  },
});

```

**Comment by janpio** [comment]
Hey, everyone that previously commented, subscribed or upvoted - or wants to have generated columns support in Prisma:
Could you please share as many examples of RLS policies as possible?

We want to figure out how much we want to or can understand and represent the "expression" of a policy in Prisma Schema Language (and use that to for example validate the expression before it is created in the database, by checking field/column names and so on). Otherwise, we could only treat them as magic SQL strings without understanding what they do, and "only" take care of migrating and introspecting them.

Thanks!

**Comment by eduhenke** [comment]
@janpio just bumping a previously mentioned one: https://github.com/prisma/prisma/issues/12735#issuecomment-1115245191

**Comment by dennemark** [comment]
> Hey, everyone that previously commented, subscribed or upvoted - or wants to have generated columns support in Prisma: Could you please share as many examples of RLS policies as possible?

@janpio 
Mine is closely related to https://github.com/prisma/prisma-client-extensions/tree/main/row-level-security
```
/**
 * Extends the base {@link PrismaClient}, associating the given information with queries made using the returned client.
 *
 * @param organisationId The ID of the organisation to associate the client with.
 * @param userId The ID of the user to associate the client with.
 * @returns An extended {@link PrismaClient} which should be used to make request on behalf of the given user.
 */
export const organisationDb = prismaDb.$extends((client) => {
  return client.$extends({
    query: {
      $allModels: {
        async $allOperations({ args, query }) {
          const userId = context.currentUser?.id
          const organisation = context.organisation as Organisation | null
          const [, , result] = await client.$transaction([
            client.$executeRaw`SELECT set_config('app.organisationId', ${organisation?.id}, TRUE)`,
            client.$executeRaw`SELECT set_config('app.userId', ${userId}, TRUE)`,
            query(args),
          ])

          return result
        },
      },
    },
  })
})

/**
 * Extends the base {@link PrismaClient}, allowing for queries which will bypass any row-level security policies which are configured to support it.
 *
 * @returns An extended {@link db Prisma Client} which can bypass any row-level security policies which support it.
 */
export const bypassDb = prismaDb.$extends((client) => {
  return client.$extends({
    query: {
      $allModels: {
        async $allOperations({ args, query }) {
          const [, result] = await client.$transaction([
            client.$executeRaw`SELECT set_config('app.bypass', 'on', TRUE)`,
            query(args),
          ])

          return result
        },
      },
    },
  })
})
```

And I added to my migration files :
```
ALTER TABLE "Project" ENABLE ROW LEVEL SECURITY;
ALTER TABLE "Project" FORCE ROW LEVEL SECURITY;
CREATE POLICY organisation ON "Project" USING ("organisationId" = current_setting('app.organisationId', TRUE)::text);
CREATE POLICY organisation_bypass ON "Project" USING (current_setting('app.bypass', TRUE)::text = 'on');
```

**Comment by andyjy** [comment]
> Could you please share as many examples of RLS policies as possible?

I use a per-user context consisting of several properties: 
- tenant_id
- workspace_id (~= organisation, accessible by multiple users)
- user_id
- boolean flag indicating admin to allow bypassing RLS for specific tables

I apply RLS policies via helper functions written in SQL, to abstract/encapsulate repeated logic. E.g.:

```sql
-- restrict access to secrets to users with the same tenant / same workspace_id in the session context
CREATE POLICY secret_rls ON secrets 
  -- secure_context is simply a separate schema holding my RLS-related functions
  -- apply_row_policies() is a custom SQL function
  USING (secure_context.apply_row_policies(
    -- first argument is an array providing the values to compare the RLS context against
    -- here, tenant_id and workspace_id are columns of the secrets table
    (ARRAY[tenant_id, workspace_id]),
    -- second argument is an array of policies to apply, in the form of SQL function names
    -- e.g. "same_tenant" is a custom SQL function to ensure the tenant_id matches
    ARRAY['secure_context.same_tenant', 'secure_context.same_workspace']
  ));
```

The policy functions look like this:

```sql
-- policy functions accept an ordered list of parameters:
-- first the values from the current transaction context,
-- then the values to use for evaluating the policy (most likely coming from the row being evaluated)
CREATE FUNCTION secure_context.same_tenant(ctx_tenant_id text, ctx_workspace_id text, ctx_user_id text, row_tenant_id text, row_workspace_id text, row_user_id text) RETURNS boolean
    LANGUAGE plpgsql
    AS $$
BEGIN
  RETURN ctx_tenant_id = row_tenant_id;
END;
$$;
```

`apply_row_policies` looks like:

```sql
CREATE FUNCTION secure_context.apply_row_policies(row_values text[], policy_functions text[]) RETURNS boolean
    LANGUAGE plpgsql
    AS $_$
DECLARE
  context            TEXT[];
  ctx_tenant_id      TEXT;
  ctx_workspace_id   TEXT;
  ctx_user_id        TEXT;

  row_tenant_id      TEXT;
  row_workspace_id   TEXT;
  row_user_id        TEXT;

  test_result  BOOLEAN;
  tests        BOOLEAN[];
  i            INT;
  f            TEXT;
BEGIN
  -- get secure context
  SELECT secure_context.get() INTO context;
  ctx_tenant_id      := context[1];
  ctx_user_id        := context[2];
  ctx_workspace_id   := context[3];

  row_tenant_id      := row_values[1];
  row_workspace_id   := row_values[2];
  row_user_id        := row_values[3];

  i := 1;
  tests[i] := true;
  FOREACH f IN ARRAY policy_functions
  LOOP
    i := i + 1;
    EXECUTE 'SELECT ' || f || '($1, $2, $3, $4, $5, $6)'
      INTO test_result
      USING ctx_tenant_id, ctx_workspace_id, ctx_user_id, row_tenant_id, row_workspace_id, row_user_id;
    tests[i] := test_result;
  END LOOP;

  RETURN true = ALL(tests);
END;
$_$;
```

`secure_context.get()` is a custom SQL function extracting the context parameters from the local transaction context, after verifying a HMAC-style hash as yet another layer of protection against fake context being injected via a SQL injection vulenrability etc: based on the approach in this article: https://www.2ndquadrant.com/en/blog/application-users-vs-row-level-security/

**Comment by jonapgar-groupby** [comment]
It's worth noting that this works:

```typescript
const [, result] = await client.$transaction([
    client.$executeRaw`SELECT set_config(${claimsSetting}, ${claims}, TRUE)`,
    query(args),
  ])
```

But it does not work when using interactive transactions...

```
const [, result] = await client.$transaction(transaction=> {
    await transaction.$executeRaw`SELECT set_config(${claimsSetting}, ${claims}, TRUE)`
    return query(args)
 })
```  

And this also doesn't work
```
const [, result] = await client.$transaction(transaction=> {
    await client.$executeRaw`SELECT set_config(${claimsSetting}, ${claims}, TRUE)`
    return query(args)
 })
```  

I hope it helps anyone facing the same issue. Took a bit to figure out what was wrong...

**Comment by icopp** [comment]
> Hey, everyone that previously commented, subscribed or upvoted - or wants to have generated columns support in Prisma: Could you please share as many examples of RLS policies as possible?

For use with Supabase, my company uses a large but ultimately pretty simple set of policies that look roughly like this:

```
-- users_profile is a 1-to-1 to the Supabase-created auth.users and contains our user metadata
CREATE POLICY users_profile_select_policy ON users_profile FOR SELECT TO authenticated USING (
    id = (
        SELECT
            auth.uid ()
    )
    OR (
        SELECT
            -- a SECURITY DEFINER function that checks internal stuff
            is_admin ()
    )
);

-- table_a is a 1-to-many table foreign keyed to users
CREATE POLICY table_a_select_policy ON table_a FOR SELECT TO authenticated
USING (
    EXISTS (
        SELECT
            1
        FROM
            users_profile
        WHERE
            table_a.user_id = users_profile.id
    )
);

-- table_b is a 1-to-many table foreign keyed to table_a
CREATE POLICY table_b_select_policy ON table_b FOR SELECT TO authenticated
USING (
    EXISTS (
        SELECT
            1
        FROM
            table_a
        WHERE
            table_a.id = table_b.table_a_id
    )
);

-- we allow users to create and update table_b entries, but only if they haven't been processed (which sets the column table_b_result)
CREATE POLICY table_b_update_policy ON table_b FOR UPDATE TO authenticated
USING (
    (
        table_b_result IS NULL
    )
    AND EXISTS (
        SELECT
            1
        FROM
            table_a
        WHERE
            table_a.id = table_b.table_a_id
    )
)
WITH CHECK (
    table_b_result IS NULL
);
```

The key thing here for efficiency, as we found out through trial and error, is the recursive use of the `EXISTS (SELECT 1 FROM ...)` clauses that resolve all the way down to one or more specific 'root' policies (in this case, the `users_profile` policy). This way, you only end up with the minor extra overhead of the query against the specifically-valid-for-this-user rows in each table corresponding with whatever your query is.

**Comment by fnimick** [comment]
There's a serious problem with the examples shown above for client extensions to support RLS: *they completely break all transaction support on the extended client*.

From what I can gather, this is occurring because the modified query is being issued on the original client, rather than the transaction client. I've created an issue to better explain: https://github.com/prisma/prisma/issues/20678

**Comment by fnimick** [comment]
I was able to piece together a hack to allow transactions to work. This creates two extended clients: one for use with transactions, and one for use without. The one used without transactions wraps each query in a transaction, and the `$transaction` method is disabled. The one used for transactions wraps the `$transaction` method to issue the query or queries used for setting the RLS setting before the transaction statements are run.

If the one used for transactions is used in a non-transactional context, no RLS claims will be set, and the queries themselves will either fail or return empty results.

```ts
export const useSupabaseRowLevelSecurity = (
  options: SupabaseRowLevelSecurityOptions = defaultSupabaseRowLevelSecurityOptions
) => {
  const name = options.name || defaultSupabaseRowLevelSecurityOptions.name;
  const claimsFn = options.claimsFn || defaultSupabaseRowLevelSecurityOptions.claimsFn;
  const claimsSetting =
    options.claimsSetting || defaultSupabaseRowLevelSecurityOptions.claimsSetting;

  return Prisma.defineExtension((client) =>
    client.$extends({
      name: name || "useSupabaseRowLevelSecurity",
      query: {
        $allModels: {
          async $allOperations({ args, query }) {
            const claims = claimsFn ? JSON.stringify(claimsFn() || {}) : "";
            try {
              const [, result] = await client.$transaction([
                client.$executeRaw`SELECT set_config(${claimsSetting}, ${claims}, TRUE)`,
                query(args),
              ]);

              return result;
            } catch (e) {
              // TODO: sentry
              if (options.logging) console.error(e);
              throw e;
            }
          },
        },
      },
      client: {
        $transaction() {
          throw new Error("Transactions not supported on this client.");
        },
      },
    })
  );
};

export const useSupabaseRowLevelSecurityForTransactions = (
  options: SupabaseRowLevelSecurityOptions = defaultSupabaseRowLevelSecurityOptions
) => {
  const name = options.name || defaultSupabaseRowLevelSecurityOptions.name;
  const claimsFn = options.claimsFn || defaultSupabaseRowLevelSecurityOptions.claimsFn;
  const claimsSetting =
    options.claimsSetting || defaultSupabaseRowLevelSecurityOptions.claimsSetting;

  return Prisma.defineExtension((client) =>
    client.$extends({
      name: name || "useSupabaseRowLevelSecurityForTransactions",
      client: {
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        $transaction(first: any, ...rest: any) {
          const claims = claimsFn ? JSON.stringify(claimsFn() || {}) : "";
          if (Array.isArray(first)) {
            // batch operation
            return client.$transaction(
              [client.$executeRaw`SELECT set_config(${claimsSetting}, ${claims}, TRUE)`, ...first],
              ...rest
            );
          }
          // function operation
          return client.$transaction(async (tx) => {
            await tx.$executeRaw`SELECT set_config(${claimsSetting}, ${claims}, TRUE)`;
            return first(tx);
          }, ...rest);
        },
      } as { $transaction: (typeof client)["$transaction"] },
    })
  );
};
```

**Comment by KozielGPC** [comment]
up

**Comment by LucianBuzzo** [comment]
For anyone looking to implement RLS in Prisma, I developed a package for doing this https://github.com/cerebruminc/yates
It uses a client extension to bootstrap and enforce RLS policies based on a provided user role.

**Comment by recurrence** [comment]
That Yates link is great, they appear to workaround all kinds of Prisma problems peering through the source.  Has anyone used that in production?  Have they encountered problems?

**Comment by LucianBuzzo** [comment]
@recurrence We use it in production at https://www.cerebrum.com/ for a number of services and it works very well. You may need to add additional indexes to your models to maintain good performance, and it's worth experimenting to make sure you get good values for wait and transaction timeouts.

**Comment by sprutner** [comment]
I am very interested in a solution for this. I am also using Supabase which sadly is reliant on RLS for security. This is a major drawback of Prisma and at some point I would need to migrate off prisma or Supabase because of this. This seems like a big want from the community. I will check out Yates although I'm trepidatious to use something so new with my database security.

**Comment by fnimick** [comment]
@sprutner  I have migrated my project away from Prisma due to the lack of native JOIN expressions causing incredibly poor performance for my use case - but when I did use Prisma, my solution was to:
- apply migrations through supabase tooling, not through prisma itself
- hand-write migrations to write the row level security policies
- set transaction variables corresponding to the user for the current request, allowing the same policies predicated on e.g. `auth.uid()` to work for both supabase API requests and prisma requests.

**Comment by chrisvander** [comment]
> const [, result] = await client.$transaction(transaction=> {
>     await transaction.$executeRaw`SELECT set_config(${claimsSetting}, ${claims}, TRUE)`
>     return query(args)
>  })

I'm confused. Why wouldn't this work?

**Comment by LucianBuzzo** [comment]
@chrisvander IIRC prisma will run the inner `$executeRaw` in a different transaction to the `query`.

**Comment by VanThanh12195** [comment]
hello everyone. I try to implement row level security for Postgresql with Prisma. is it a correct to do this? I need to get a current_user_id from session (after user loged in) and pass that id to the Prisma extend query. since I need to do that async await to get the id. I am not sure this is a correct way to instantiate PrismaClient and re-use it through the entire application.

`import { PrismaClient } from "@prisma/client";

const getPrismaClient = async () => {
  const supabaseClient = getSupabaseServer();

  const { id } = (await supabaseClient.auth.getSession()).data.session.user;

  const prisma = new PrismaClient().$extends({
    query: {
      $allModels: {
        async $allOperations({ args, query }) {
          const [, result] = await prisma.$transaction([
            prisma.$executeRaw`SELECT set_config('app.current_user_id', ${id}, TRUE)`,
            query(args),
          ]);
          return result;
        },
      },
    },
  });

  return prisma;
};

const globalForPrisma = globalThis;

const prisma =
  globalForPrisma.prisma ??
  (async () => {
    await getPrismaClient();
  })();

export default prisma;

if (process.env.NODE_ENV !== "production") globalForPrisma.prisma = prisma;`

**Comment by VanThanh12195** [comment]
Hello everyone. I try to implement row level security for Postgresql with Prisma. Is it a correct to do this? I need to get a current_user_id from session (after user loged in) and pass that id to the Prisma extend query. since I need to do that async await to get the id. I am not sure this is a correct way to instantiate PrismaClient and re-use it through the entire application. Thank you all.


```javascript
import { PrismaClient } from "@prisma/client";

const getPrismaClient = async () => {

  const { id } =await getCurrentUserId(); //this is an example function to get current user Id.

  const prisma = new PrismaClient().$extends({
    query: {
      $allModels: {
        async $allOperations({ args, query }) {
          const [, result] = await prisma.$transaction([
            prisma.$executeRaw`SELECT set_config('app.current_user_id', ${id}, TRUE)`,
            query(args),
          ]);
          return result;
        },
      },
    },
  });

  return prisma;
};

const globalForPrisma = globalThis;

const prisma =
  globalForPrisma.prisma ??
  (async () => {
    await getPrismaClient();
  })();

export default prisma;

if (process.env.NODE_ENV !== "production") globalForPrisma.prisma = prisma;
```

**Comment by chrisvander** [comment]
By default, on every request, sounds like a breaking change and high overhead. Should be opt-in.

**Comment by shoooe** [comment]
> By default, on every request, sounds like a breaking change and high overhead. Should be opt-in.

Oh I'm not talking at the Prisma level. Rather at the web server level (assuming you are using Prisma in a Web API server like I do).

**Comment by wmtrinu** [comment]
> @sprutner I have migrated my project away from Prisma due to the lack of native JOIN expressions causing incredibly poor performance for my use case - but when I did use Prisma, my solution was to:
> 
> * apply migrations through supabase tooling, not through prisma itself
> * hand-write migrations to write the row level security policies
> * set transaction variables corresponding to the user for the current request, allowing the same policies predicated on e.g. `auth.uid()` to work for both supabase API requests and prisma requests.

what are you using at the moment?

**Comment by ciekawy** [comment]
is this a blocker to use prisam with supabase RLS? or should it be possible to use proposed workarounds or `yates`?

**Comment by LazerJesus** [comment]
this would be such a nice value add for prisma and keep it relevant in the context of supabase.

**Comment by fnimick** [comment]
> > @sprutner I have migrated my project away from Prisma due to the lack of native JOIN expressions causing incredibly poor performance for my use case - but when I did use Prisma, my solution was to:
> > 
> > * apply migrations through supabase tooling, not through prisma itself
> > * hand-write migrations to write the row level security policies
> > * set transaction variables corresponding to the user for the current request, allowing the same policies predicated on e.g. `auth.uid()` to work for both supabase API requests and prisma requests.
> 
> what are you using at the moment?

I migrated to Drizzle. Here's my solution for constructing an RLS client based on the current user info in the session: https://github.com/drizzle-team/drizzle-orm/issues/594#issuecomment-1871432968

**Comment by kiptoomm** [comment]
Has anyone here considered the Zenstack solution offered by @jiashengguo? https://github.com/prisma/prisma/issues/5128#issuecomment-1466207852

I'm curious to hear about their experience and performance considerations, if any

**Comment by grokpot** [comment]
@kiptoomm I just went down a rabbit hole on this.
I'm using Vue + Nuxt + Supabase + Supabase GraphQL + Prisma
Supabase RLS with Prisma kept getting stickier. Zenstack looks cool (nice job @jiashengguo), but:
1. It makes me feel uncomfortable to use Zenstack, because it's _another_ layer on top of Prisma which is _another_ layer on the DB. Zenstack should really just be native Prisma (sorry if that's a naive take)
2. Regardless, [I still need to use supabase migrations to enable RLS](https://zenstack.dev/docs/guides/supabase-security#3-how-to-prevent-direct-supabase-api-requests) (you could argue "just don't use supabase!" but that's a different discussion.
3. If I use Zenstack, [I need to "enhance" my prisma client](https://zenstack.dev/docs/guides/authentication/supabase#creating-an-enhanced-prisma-client). This is more [straightforward with NEXT](https://supabase.com/docs/guides/auth/auth-helpers), but [Nuxt/supabase](https://supabase.nuxtjs.org/) has it's own library, which makes this more difficult (especially considering I'd probably need to write a Nitro plugin, which is a new-ish thing and not well documented)
4. I'm using Supabase GraphQL, which means adding permissions to Zenstack wouldn't affect my GraphQL calls 😆 

I ended up just removing Prisma and going with straight Supabase.
Cons: migrations are more Django-ish with Prisma (rollbacks), I like the declarative schema file of Prisma
Pros: layer of complexity removed, graphQL and REST abide by the same rules, don't have to install another layer (Zenstack), maintain one migration folder instead of two, get better at SQL

Hope that helps someone out there.

**Comment by kiptoomm** [comment]
> @kiptoomm I just went down a rabbit hole on this. I'm using Vue + Nuxt + Supabase + Supabase GraphQL + Prisma Supabase RLS with Prisma kept getting stickier. Zenstack looks cool (nice job @jiashengguo), but:
> 
> 1. It makes me feel uncomfortable to use Zenstack, because it's _another_ layer on top of Prisma which is _another_ layer on the DB. Zenstack should really just be native Prisma (sorry if that's a naive take)
> 2. Regardless, [I still need to use supabase migrations to enable RLS](https://zenstack.dev/docs/guides/supabase-security#3-how-to-prevent-direct-supabase-api-requests) (you could argue "just don't use supabase!" but that's a different discussion.
> 3. If I use Zenstack, [I need to "enhance" my prisma client](https://zenstack.dev/docs/guides/authentication/supabase#creating-an-enhanced-prisma-client). This is more [straightforward with NEXT](https://supabase.com/docs/guides/auth/auth-helpers), but [Nuxt/supabase](https://supabase.nuxtjs.org/) has it's own library, which makes this more difficult (especially considering I'd probably need to write a Nitro plugin, which is a new-ish thing and not well documented)
> 4. I'm using Supabase GraphQL, which means adding permissions to Zenstack wouldn't affect my GraphQL calls 😆
> 
> I ended up just removing Prisma and going with straight Supabase. Cons: migrations are more Django-ish with Prisma (rollbacks), I like the declarative schema file of Prisma Pros: layer of complexity removed, graphQL and REST abide by the same rules, don't have to install another layer (Zenstack), maintain one migration folder instead of two, get better at SQL
> 
> Hope that helps someone out there.

@grokpot thanks for the insightful reply. I have the same concerns about 'layering' w.r.t Zenstack, but will still give it a try because the declarative approach looks elegant (the primary feature that drew me to Prisma in the first place). On the other hand, Prisma's [solution](https://supabase.com/docs/guides/auth/row-level-security) utilizing client extensions seems too convoluted to me. Per the [example project](https://github.com/prisma/prisma-client-extensions/tree/main/row-level-security):
* We still have to write some SQL _anyway_ (yet the ORM should ideally remove this need for those of us who aren't comfortable putting app logic into the database). See steps 1-6
* The solution is [not production-ready](https://github.com/prisma/prisma-client-extensions/tree/main/row-level-security#caveats):
> This extension is provided as an example only. It is not intended to be used in production environments.

👆 makes me wonder what the value/goal of these extensions are

**Comment by jason-curtis** [comment]
@LucianBuzzo taking a deep look at your [Yates ](https://github.com/cerebruminc/yates) package. Seems like a great concept.

The most surprising thing to me is that you chose to create RLS roles and grants on startup (rather than using migrations for instance). I love the promise that I don't have to maintain migrations for RLS, but can you comment on the tradeoffs there? Are there overhead costs of doing all of this on startup? Does a new set of roles and grants get created on every instance or is there some kind of diffing/updating approach?

**Comment by LucianBuzzo** [comment]
@jason-curtis When creating Yates, I really wanted to make it as low-config as possible. Setting up user permissions can be tricky, so being able to iterate quickly and make changes without having to go through multiple setup/configuration steps is very convenient. Adding extra steps also creates additional points of failure that can create confusion and frustration for the end user. 

On startup, Yates will compute the expressions for each of the abilities defined, and check them against the database to see if they need to be created or updated. They will only be changed if necessary. This compute and check process adds some overhead to the server startup time, on our production system at Cerebrum, we have 148* default abilities and 133 custom abilities and it takes ~2.5s to get through the setup process for Yates. This delay is completely acceptable for us as we are using a K8s cluster for deployment and waiting for health checks to complete before switching the container live.

The startup time could certainly be an issue if you are running Prisma in a serverless/edge-function style and have a lot of models and abilities, but as this is not our architecture I've not spent much time improving startup performance. This conversation has definitely got my mind turning on what you could do to improve startup performance, and I think there is a lot of headroom to get the time down.

_*37 models each with 4 default CRUD abilties_

**Comment by jason-curtis** [comment]
@LucianBuzzo thanks for the insight. The app I'm currently working on is serverless (likely scaling-to-zero) so reducing startup time is important. Would love to hear more about this possibility without spamming this thread, so I opened a separate discussion: https://github.com/cerebruminc/yates/issues/90

**Comment by sbue** [comment]
@LucianBuzzo, thinking of implementing this for Supabase RLS

Would be awesome if the repo had some kind of example branch for working with Supabase

**Comment by LucianBuzzo** [comment]
@sbue Feel free to open an issue about supabase support on the Yates repo and we can focus discussion there 👍

**Comment by janpio** [comment]
Hey, a while ago [I had written this](https://github.com/prisma/prisma/issues/12735#issuecomment-1492658385):

> Hey, everyone that previously commented, subscribed or upvoted - or wants to have generated columns support in Prisma: Could you please share as many examples of RLS policies as possible?
> 
> We want to figure out how much we want to or can understand and represent the "expression" of a policy in Prisma Schema Language (and use that to for example validate the expression before it is created in the database, by checking field/column names and so on). Otherwise, we could only treat them as magic SQL strings without understanding what they do, and "only" take care of migrating and introspecting them.

In the comments after, some of you thankfully provided examples for specific `CREATE POLICY` statement that you use. Thanks! 🙇 

Now it is time to expand that a bit, and I would love to have **full database schemas** that include both the database tables and the security policies that apply to these tables. If _you_ have a database that uses PostgreSQL Row Level Security, I would love if you could share that with me via schemas@prisma.io. 

We will use that to further understand the details on how we could represent these policies in Prisma Schema Language, to enable migration and introspection of RLS and its policies. Thank you.

**Comment by davidgomes** [comment]
I took at stab at how the Prisma ORM language could be expanded to support PG RLS. I put together an example in this Gist: https://gist.github.com/davidgomes/cfa4607d6d004b262802dedc365541f2.

This is a bit of a 1-1 mapping between PG RLS Syntax and new Prisma syntax. The reason why I went with that approach as opposed to [something like what ZenStack does](https://zenstack.dev/docs/reference/zmodel-language#allow), is because ZenStack works very differently from Prisma. ZenStack generates an API server which writes queries according to authorization rules. But I would expect Prisma at its core level to not depend on some proxy and instead just generate PG RLS Policies natively.

Curious to know if others have any input on this API!

**Comment by chrisvander** [comment]
@davidgomes I like your gist but my only thought is that nested lookups don't really appear possible. I have an app, let's say I have a User who belongs to an Organization and has access to Projects. While the Organization may have many Projects, the User would only have access to a subset. So, a context would be something like 
```
context {
  user User
  organization Organization
  projects Project[]
}
```
^ or just the corresponding IDs.

The nesting case would be something like: I have a Document which contains ContentBlock models. In this case, the ContentBlock model doesn't care about the User, Organization, or Project, it only cares about the Document it's related to. Access to a ContentBlock depends on access to the Document. So a security policy for the ContentBlock would need to be something like:
```
@@security({
  name: 'access_content_block'
  read: { document: { project: { in: context.projects } } }
})
```

I think that bit is essential to an RLS implementation, so that you don't need to add columns to a table where there isn't an application need.

**Comment by sorenbs** [comment]
@chrisvander ,

What do you think about implementing something simple like David suggests for 80% of the use cases, and then have a catch-all mechanism for more complex use cases like what you describe?

The catch-all would simply read an arbitrary RLS statement from a `.sql` file in your prisma folder:


With this in your schema

```
@@security({
  name: 'access_content_block'
})
```

Prisma will expect a file located at `./prisma/rls/access_content_block.sql`

**Comment by chrisvander** [comment]
@sorenbs I could see that working well for complex use cases, absolutely. Being able to reference a SQL file (which I assume contains a policy definition?) would ensure that whatever model you use, there will be ways to implement more complex use cases.

Having the option to reference a column directly (like "userId") or be able to reference a path to a reference (like "x.y.userId") still represents a "simple use case" in my mind. Anything beyond looking up a single column through relations or directly would represent the complex case. Being able to work with columns on relations, for my code base at least, would eliminate 90% of any arbitrary `.sql` statements in a `./prisma/rls` folder.

But, where Prisma is right now, I certainly wouldn't complain about a basic but Prisma-managed RLS implementation with a complex use case mechanism, and have something like the above come later.

**Comment by jiashengguo** [comment]
> @davidgomes I like your gist but my only thought is that nested lookups don't really appear possible. I have an app, let's say I have a User who belongs to an Organization and has access to Projects. While the Organization may have many Projects, the User would only have access to a subset. So, a context would be something like
> 
> ```
> context {
>   user User
>   organization Organization
>   projects Project[]
> }
> ```
> 
> ^ or just the corresponding IDs.
> 
> The nesting case would be something like: I have a Document which contains ContentBlock models. In this case, the ContentBlock model doesn't care about the User, Organization, or Project, it only cares about the Document it's related to. Access to a ContentBlock depends on access to the Document. So a security policy for the ContentBlock would need to be something like:
> 
> ```
> @@security({
>   name: 'access_content_block'
>   read: { document: { project: { in: context.projects } } }
> })
> ```
> 
> I think that bit is essential to an RLS implementation, so that you don't need to add columns to a table where there isn't an application need.
> 
> I will acknowledge, I don't know how to make it a 1-to-1 mapping. Being able to generate some nested JOIN seems important, though.

I think @chrisvander has made a great point. The nested lookup ability is essential, both in terms of value and implementation.

**Comment by davidgomes** [comment]
@ChristianJacobsen I think you bring up a great point, and I especially like that you re-surfaced the `context` part of this. It might be a good idea for Prisma to support _typing_ the connection context (i.e., the set of local variables that should be set on every database connection). And I would also also hope that the Prisma client is able to validate and reject database connections that **do not** meet the right context. **A lot of people prefer to use RLS with a rich context such as a JWT object instead of relying on Postgres roles**.

So, maybe Prisma could indeed have a `context` definer, as has been discussed before on this issue:

```
context {
    id         String
    role      String
}
```

And for JWT contexts, we could have something like:

```
context {
  @jwt("id")     id String
  @jwt("role") role String
}
```

And then, these could be used in different places, perhaps even inside `dbgenerated`:

```
@@rls({
  name: "only admins can manage books",
  for: ["create", "update", "delete"],
  check: dbgenerated(`${context.role} = "admin"`),
})
```

And for nested lookups, we can then do the following (imagine that the application sets a set of project IDs in an array in the database connection's local variables, which is a common pattern for RLS users):

```
context {
  userId    String
  projects String[]
  role String
}

...
model Review {
  ...
  authorId        Int      @id @default(autoincrement())

  @@rls({
    name: "either admins or review authors can edit reviews",
    for: ["create", "update", "delete"],
    check: `dbgenerated(${context.role} = "admin") OR dbgenerated(${context.id} = authorId)`
  })

  ...
}
```

> @chrisvander ,
> 
> What do you think about implementing something simple like David suggests for 80% of the use cases, and then have a catch-all mechanism for more complex use cases like what you describe?
> 
> The catch-all would simply read an arbitrary RLS statement from a `.sql` file in your prisma folder:
> 
> With this in your schema
> 
> ```
> @@security({
>   name: 'access_content_block'
> })
> ```
> 
> Prisma will expect a file located at `./prisma/rls/access_content_block.sql`

I really like this! I see it as a fallback utility for the 10-20% of more complicated RLS policies.

**Comment by chrisvander** [comment]
> It might be a good idea for Prisma to support typing the connection context (i.e., the set of local variables that should be set on every database connection). And I would also also hope that the Prisma client is able to validate and reject database connections that do not meet the right context. **A lot of people prefer to use RLS with a rich context such as a JWT object instead of relying on Postgres roles.**

It's funny that you bring this up, I hadn't even thought of it, but that's how my basic RLS setup is wired up anyway.

You would share a Prisma Client instance between multiple requests and so you'd want to pass a different context for each database call. Maybe via a `$extends` middleware or something similar.

**Comment by dennemark** [comment]
It might good to revisit some of the RLS client extensions and only allow interactive transactions, since sequential transactions can easily break and won't roll back when used with client extensions! I wrote a little guide: https://github.com/prisma/prisma/discussions/25034 . Haven't tested it for RLS, but for my other extension ( i am going to ditch RLS for my prisma-extension-casl which integrates casl.js authorization in my client even for nested queries - still WIP).

**Comment by danielrose7** [comment]
We had some issues with other solutions--including that we rely on sequential batch transactions rather than just interactive transactions to reduce database trips.

Here is a little gist of a solution that extends prisma's client: https://gist.github.com/danielrose7/a0c6a98e7de5f63dffb86ec05d34dab9

This allows:
- setting RLS before operations on prisma models via allModels
- setting RLS before operations using raw sql operations ($queryRaw, $queryRawUnsafe, $executeRaw, $executeRawUnsafe)
- interactive transactions (`$transaction(async (transaction) => { await ... })`)
- batch sequential transactions (`$transaction([op1, op2])`)

It seems helpful to note that extending the prisma client in this way doesn't necessarily allow for extensions that come later in the chain to "just work". I got around this by accepting an optional 'extensions' argument that can apply arbitary extensions before this 'bats last'

If need to rework for your use case, I recommend working with the jest test and your environment rather than debugging directly. Can get tricky in there!

If opt to use something like the internal '_createItxClient', I recommend leaving a watermark on the existing transaction so that you don't end up setting the RLS context multiple times in the same transaction.

Otherwise you're liable to set 1x per operation

```ts
// extension snippet of __internalParams -> transaction
$allModels: {
  async $allOperations({ args, query, __internalParams, ...rest }) {
    const existingTxn = __internalParams?.transaction

    if (existingTxn && !existingTxn._rlsApplied) {
      const transactionClient = (prismaExtension as any)._createItxClient(existingTxn);
      await transactionClient.$executeRaw(setRlsSql); //  set your RLS here!
      existingTxn._rlsApplied = true // leave a watermark for the next operation to avoid this if block
    }

   // ... normal operations here
...
}
```

**Comment by zhzhang** [comment]
Rather than the `set_config` approach, is there any inherent issue with passing the user id via connection query params?
```
export function createAuthenticatedPrismaClient(userId: string): PrismaClient {
  const databaseUrl = `${safeEnv.databaseUrl}?options=-c%20app.meta.client_id%3D${encodeURIComponent(userId)}`;
  return new PrismaClient({
    datasourceUrl: databaseUrl,
  });
}
```
and
```
SELECT current_setting('app.meta.client_id')
```
to use the value in RLS context later on.
Are there any inherent issues with this approach?

**Comment by LucianBuzzo** [comment]
@zhzhang The issue you will face here is that you will need a new client for each of the distinct users you have. If this is tolerable, then embedding the `userId` in the connection string would be fine.

**Comment by zhzhang** [comment]
@LucianBuzzo, got it. I'm a bit of a novice w.r.t. postgres - what are the implications of having to make a new client per user (or even per query)? Does it come with significant performance tradeoffs?

**Comment by LucianBuzzo** [comment]
@zhzhang Have a separate client per user will negatively impacts on query speed and your application:
- *Connection limits*: you run the risk of exhausting available connections to your database
- *Startup time*: Starting a new client adds latency as it goes through it's startup process and initializes a connection
- *Memory issues*: Your increasing the amount of active memory consumption, and also adding complexity to your system as you need to handle stale connections and correctly dispose of old client
